src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs

```diff
@@ -23,22 +27,80 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
     /// </summary>
     [ExportRoslynLanguagesLspRequestHandlerProvider(typeof(SemanticTokensRangeHandler)), Shared]
     [Method(Methods.TextDocumentSemanticTokensRangeName)]
-    internal class SemanticTokensRangeHandler : AbstractStatelessRequestHandler<LSP.SemanticTokensRangeParams, LSP.SemanticTokens>
+    internal class SemanticTokensRangeHandler : AbstractStatelessRequestHandler<LSP.SemanticTokensRangeParams, LSP.SemanticTokens>, IDisposable
     {
         private readonly IGlobalOptionService _globalOptions;
-
-        // Lock to guard _projectToCompilation dictionary
-        private readonly object _lock = new();
-        private readonly Dictionary<ProjectId, Task<Compilation?>> _projectIdToCompilation = new();
+        private readonly IAsynchronousOperationListener _asyncListener;
 
         public override bool MutatesSolutionState => false;
         public override bool RequiresLSPSolution => true;
 
+        private readonly object _gate = new();
+        private readonly Dictionary<ProjectId, CompilationAvailableEventSource> _projectIdToEventSource = new();
```

Comment:
could this be a pair of last checksum and compilationavailable event source in one map?